File: src/main/java/ru/big/survey/service/Conditions.java
package ru.big.survey.service;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import tools.jackson.databind.JsonNode;

/**
 * Интерпретатор условий показа элементов анкеты — порт conditions.js из v1 (без eval).
 * Строка вида: Город == "Москва" || (Статус != "Другое" && !Согласие); операторы == != > >= < <=, || &&, !, скобки,
 * литералы "строка"/'строка', числа, true/false; идентификаторы — имена полей ответов.
 * Структурированная форма: {field, op, value} | {any:[…]} | {all:[…]} | {not: …}.
 * Пустое условие → true. Синтаксическая ошибка → true (потерять поле хуже, чем показать лишнее).
 */
public final class Conditions {

    private Conditions() {
    }

    private static final Map<String, Node> CACHE = new ConcurrentHashMap<>();

    /** Видимость элемента (компонента или поля) по его свойству conditions при данных ответах. */
    public static boolean isVisible(JsonNode item, JsonNode answers) {
        if (item == null || !item.isObject()) {
            return true;
        }
        return evaluate(item.get("conditions"), answers);
    }

    public static boolean evaluate(JsonNode condition, JsonNode answers) {
        Node ast;
        try {
            ast = compile(condition);
        } catch (RuntimeException e) {
            return true;
        }
        if (ast == null) {
            return true;
        }
        return truthy(ast.eval(answers));
    }

    static Node compile(JsonNode condition) {
        if (condition == null || condition.isNull() || condition.isMissingNode()) {
            return null;
        }
        if (condition.isString()) {
            String text = condition.stringValue();
            if (text.isBlank()) {
                return null;
            }
            return CACHE.computeIfAbsent(text, Conditions::parse);
        }
        if (condition.isObject()) {
            return fromObject(condition);
        }
        throw new IllegalArgumentException("Условие должно быть строкой или объектом");
    }

    // ---------- AST ----------

    sealed interface Node permits Field, Literal, Not, And, Or, Compare {
        Object eval(JsonNode answers);
    }

    record Field(String name) implements Node {
        public Object eval(JsonNode answers) {
            if (answers == null) {
                return null;
            }
            JsonNode v = answers.get(name);
            return v == null || v.isNull() ? null : v;
        }
    }

    record Literal(Object value) implements Node {
        public Object eval(JsonNode answers) { return value; }
    }

    record Not(Node operand) implements Node {
        public Object eval(JsonNode answers) { return !truthy(operand.eval(answers)); }
    }

    record And(Node left, Node right) implements Node {
        public Object eval(JsonNode answers) { return truthy(left.eval(answers)) && truthy(right.eval(answers)); }
    }

    record Or(Node left, Node right) implements Node {
        public Object eval(JsonNode answers) { return truthy(left.eval(answers)) || truthy(right.eval(answers)); }
    }

    record Compare(String op, Node left, Node right) implements Node {
        public Object eval(JsonNode answers) {
            Object l = left.eval(answers);
            Object r = right.eval(answers);
            switch (op) {
                case "==": return looseEquals(l, r);
                case "!=": return !looseEquals(l, r);
                default: {
                    Double ln = numeric(l);
                    Double rn = numeric(r);
                    if (ln == null || rn == null) {
                        return false;
                    }
                    return switch (op) {
                        case ">" -> ln > rn;
                        case ">=" -> ln >= rn;
                        case "<" -> ln < rn;
                        default -> ln <= rn;
                    };
                }
            }
        }
    }

    // ---------- семантика значений (как в v1) ----------

    static boolean truthy(Object v) {
        if (v == null) {
            return false;
        }
        if (v instanceof Boolean b) {
            return b;
        }
        if (v instanceof Number n) {
            return n.doubleValue() != 0;
        }
        if (v instanceof String s) {
            return !s.isEmpty() && !s.equalsIgnoreCase("false");
        }
        if (v instanceof JsonNode n) {
            if (n.isNull() || n.isMissingNode()) {
                return false;
            }
            if (n.isArray()) {
                return n.size() > 0;
            }
            if (n.isBoolean()) {
                return n.asBoolean();
            }
            if (n.isNumber()) {
                return n.asDouble() != 0;
            }
            if (n.isString()) {
                String s = n.stringValue();
                return !s.isEmpty() && !s.equalsIgnoreCase("false");
            }
            return true;
        }
        return true;
    }

    private static boolean looseEquals(Object a, Object b) {
        if (a instanceof JsonNode an && an.isArray()) {
            for (JsonNode x : an) {
                if (looseEquals(x, b)) {
                    return true;
                }
            }
            return false;
        }
        if (b instanceof JsonNode bn && bn.isArray()) {
            for (JsonNode x : bn) {
                if (looseEquals(a, x)) {
                    return true;
                }
            }
            return false;
        }
        boolean aBool = a instanceof Boolean || (a instanceof JsonNode n && n.isBoolean());
        boolean bBool = b instanceof Boolean || (b instanceof JsonNode n && n.isBoolean());
        if (aBool || bBool) {
            return truthy(a) == truthy(b);
        }
        return asString(a).equals(asString(b));
    }

    private static String asString(Object v) {
        if (v == null) {
            return "";
        }
        if (v instanceof JsonNode n) {
            if (n.isNull() || n.isMissingNode()) {
                return "";
            }
            if (n.isString()) {
                return n.stringValue();
            }
            if (n.isNumber()) {
                double d = n.asDouble();
                return d == Math.rint(d) && !Double.isInfinite(d) ? String.valueOf((long) d) : String.valueOf(d);
            }
            return n.toString();
        }
        if (v instanceof Double d) {
            return d == Math.rint(d) && !Double.isInfinite(d) ? String.valueOf(d.longValue()) : String.valueOf(d);
        }
        return String.valueOf(v);
    }

    private static Double numeric(Object v) {
        if (v == null) {
            return null;
        }
        if (v instanceof Number n) {
            return n.doubleValue();
        }
        String s = v instanceof JsonNode n
                ? (n.isNumber() ? String.valueOf(n.asDouble()) : (n.isString() ? n.stringValue() : ""))
                : String.valueOf(v);
        try {
            return Double.parseDouble(s.trim().replace(',', '.'));
        } catch (NumberFormatException e) {
            return null;
        }
    }

    // ---------- структурированная форма ----------

    private static Node fromObject(JsonNode cond) {
        if (cond.isString()) {
            return compile(cond);
        }
        if (!cond.isObject()) {
            throw new IllegalArgumentException("Неизвестная форма условия");
        }
        if (cond.get("any") != null && cond.get("any").isArray()) {
            Node acc = null;
            for (JsonNode c : cond.get("any")) {
                Node n = fromObject(c);
                acc = acc == null ? n : new Or(acc, n);
            }
            return acc == null ? new Literal(true) : acc;
        }
        if (cond.get("all") != null && cond.get("all").isArray()) {
            Node acc = null;
            for (JsonNode c : cond.get("all")) {
                Node n = fromObject(c);
                acc = acc == null ? n : new And(acc, n);
            }
            return acc == null ? new Literal(true) : acc;
        }
        if (cond.get("not") != null) {
            return new Not(fromObject(cond.get("not")));
        }
        if (cond.get("field") != null && cond.get("field").isString()) {
            String op = cond.get("op") == null ? "==" : cond.get("op").asString();
            if (!List.of("==", "!=", ">", ">=", "<", "<=").contains(op)) {
                throw new IllegalArgumentException("Неизвестный оператор " + op);
            }
            JsonNode value = cond.get("value");
            Object literal = value == null || value.isNull() ? null
                    : value.isBoolean() ? value.asBoolean()
                    : value.isNumber() ? value.asDouble()
                    : value.isString() ? value.stringValue()
                    : value;
            return new Compare(op, new Field(cond.get("field").stringValue()), new Literal(literal));
        }
        throw new IllegalArgumentException("Неизвестная форма условия");
    }

    // ---------- строковая форма: токенизатор + рекурсивный спуск ----------

    private record Token(String type, Object value) {}

    static List<Token> tokenize(String src) {
        List<Token> tokens = new ArrayList<>();
        int i = 0;
        int n = src.length();
        while (i < n) {
            char ch = src.charAt(i);
            if (ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r') {
                i++;
                continue;
            }
            if (ch == '"' || ch == '\'') {
                char quote = ch;
                int j = i + 1;
                StringBuilder value = new StringBuilder();
                while (j < n && src.charAt(j) != quote) {
                    if (src.charAt(j) == '\\' && j + 1 < n) {
                        value.append(src.charAt(j + 1));
                        j += 2;
                    } else {
                        value.append(src.charAt(j));
                        j++;
                    }
                }
                if (j >= n) {
                    throw new IllegalArgumentException("Незакрытая строка с позиции " + i);
                }
                tokens.add(new Token("string", value.toString()));
                i = j + 1;
                continue;
            }
            if (Character.isDigit(ch)) {
                int j = i;
                while (j < n && (Character.isDigit(src.charAt(j)) || src.charAt(j) == '.')) {
                    j++;
                }
                tokens.add(new Token("number", Double.parseDouble(src.substring(i, j))));
                i = j;
                continue;
            }
            if (isIdentStart(ch)) {
                int j = i;
                while (j < n && isIdentChar(src.charAt(j))) {
                    j++;
                }
                String word = src.substring(i, j);
                if (word.equals("true") || word.equals("false")) {
                    tokens.add(new Token("boolean", word.equals("true")));
                } else {
                    tokens.add(new Token("ident", word));
                }
                i = j;
                continue;
            }
            String three = src.substring(i, Math.min(n, i + 3));
            String two = src.substring(i, Math.min(n, i + 2));
            if (three.equals("===") || three.equals("!==")) {
                tokens.add(new Token("op", three.substring(0, 2)));
                i += 3;
                continue;
            }
            if (List.of("==", "!=", ">=", "<=", "||", "&&").contains(two)) {
                tokens.add(new Token("op", two));
                i += 2;
                continue;
            }
            if (ch == '>' || ch == '<' || ch == '!' || ch == '(' || ch == ')') {
                tokens.add(new Token("op", String.valueOf(ch)));
                i++;
                continue;
            }
            throw new IllegalArgumentException("Неожиданный символ «" + ch + "» в позиции " + i);
        }
        return tokens;
    }

    private static boolean isIdentStart(char c) {
        return Character.isLetter(c) || c == '_';
    }

    private static boolean isIdentChar(char c) {
        return Character.isLetterOrDigit(c) || c == '_';
    }

    static Node parse(String src) {
        Parser p = new Parser(tokenize(src));
        Node ast = p.or();
        if (p.pos < p.tokens.size()) {
            throw new IllegalArgumentException("Лишний токен после конца условия");
        }
        return ast;
    }

    private static final class Parser {
        final List<Token> tokens;
        int pos;

        Parser(List<Token> tokens) {
            this.tokens = tokens;
        }

        Token peek() { return pos < tokens.size() ? tokens.get(pos) : null; }
        Token next() { return tokens.get(pos++); }

        boolean isOp(String v) {
            Token t = peek();
            return t != null && t.type().equals("op") && t.value().equals(v);
        }

        Node primary() {
            Token t = peek();
            if (t == null) {
                throw new IllegalArgumentException("Неожиданный конец условия");
            }
            next();
            switch (t.type()) {
                case "ident": return new Field((String) t.value());
                case "string", "number", "boolean": return new Literal(t.value());
                default:
                    if (t.value().equals("(")) {
                        Node e = or();
                        if (!isOp(")")) {
                            throw new IllegalArgumentException("Ожидалась закрывающая скобка");
                        }
                        next();
                        return e;
                    }
                    throw new IllegalArgumentException("Неожиданный токен «" + t.value() + "»");
            }
        }

        Node compare() {
            Node left = primary();
            Token t = peek();
            if (t != null && t.type().equals("op") && List.of("==", "!=", ">", ">=", "<", "<=").contains((String) t.value())) {
                String op = (String) next().value();
                Node right = primary();
                return new Compare(op, left, right);
            }
            return left;
        }

        Node unary() {
            if (isOp("!")) {
                next();
                return new Not(unary());
            }
            return compare();
        }

        Node and() {
            Node left = unary();
            while (isOp("&&")) {
                next();
                left = new And(left, unary());
            }
            return left;
        }

        Node or() {
            Node left = and();
            while (isOp("||")) {
                next();
                left = new Or(left, and());
            }
            return left;
        }
    }
}

File: frontend/src/lib/conditions.ts
/**
 * Безопасный интерпретатор условий показа элементов анкеты (порт conditions.js из v1, без eval).
 *
 * Строка:  Город == "Москва" || (Статус != "Другое" && !Согласие)
 *   операторы == != === !== > >= < <=, || &&, !, скобки; литералы "строка"/'строка', числа, true/false;
 *   идентификаторы — имена полей ответов (русские/латинские буквы, цифры, _).
 * Объект:  { field, op, value } | { any: [...] } | { all: [...] } | { not: ... }
 * Пустое условие → виден. Синтаксическая ошибка → виден (потерять поле хуже, чем показать лишнее).
 */

export type Answers = Record<string, unknown>;
export type Condition = string | ConditionObject | null | undefined;
export interface ConditionObject {
  field?: string;
  op?: string;
  value?: unknown;
  any?: Condition[];
  all?: Condition[];
  not?: Condition;
}

type Node =
  | { type: "field"; name: string }
  | { type: "literal"; value: unknown }
  | { type: "not"; operand: Node }
  | { type: "and"; left: Node; right: Node }
  | { type: "or"; left: Node; right: Node }
  | { type: "compare"; op: string; left: Node; right: Node };

type Token = { type: "string" | "number" | "boolean" | "ident" | "op"; value: unknown };

const IDENT_START = /[A-Za-zА-Яа-яЁё_]/;
const IDENT_CHAR = /[A-Za-zА-Яа-яЁё0-9_]/;
const DIGIT = /[0-9]/;
const COMPARE_OPS = ["==", "!=", ">", ">=", "<", "<="];

export function tokenize(src: string): Token[] {
  const tokens: Token[] = [];
  const s = String(src);
  let i = 0;
  while (i < s.length) {
    const ch = s[i];
    if (ch === " " || ch === "\t" || ch === "\n" || ch === "\r") {
      i++;
      continue;
    }
    if (ch === '"' || ch === "'") {
      const quote = ch;
      let j = i + 1;
      let value = "";
      while (j < s.length && s[j] !== quote) {
        if (s[j] === "\\" && j + 1 < s.length) {
          value += s[j + 1];
          j += 2;
        } else {
          value += s[j];
          j++;
        }
      }
      if (j >= s.length) throw new SyntaxError(`Незакрытая строка начиная с позиции ${i}`);
      tokens.push({ type: "string", value });
      i = j + 1;
      continue;
    }
    if (DIGIT.test(ch)) {
      let j = i;
      while (j < s.length && (DIGIT.test(s[j]) || s[j] === ".")) j++;
      tokens.push({ type: "number", value: Number(s.slice(i, j)) });
      i = j;
      continue;
    }
    if (IDENT_START.test(ch)) {
      let j = i;
      while (j < s.length && IDENT_CHAR.test(s[j])) j++;
      const word = s.slice(i, j);
      if (word === "true" || word === "false") tokens.push({ type: "boolean", value: word === "true" });
      else tokens.push({ type: "ident", value: word });
      i = j;
      continue;
    }
    const three = s.slice(i, i + 3);
    const two = s.slice(i, i + 2);
    if (three === "===" || three === "!==") {
      tokens.push({ type: "op", value: three.slice(0, 2) });
      i += 3;
      continue;
    }
    if (["==", "!=", ">=", "<=", "||", "&&"].includes(two)) {
      tokens.push({ type: "op", value: two });
      i += 2;
      continue;
    }
    if (ch === ">" || ch === "<" || ch === "!" || ch === "(" || ch === ")") {
      tokens.push({ type: "op", value: ch });
      i++;
      continue;
    }
    throw new SyntaxError(`Неожиданный символ «${ch}» в позиции ${i}`);
  }
  return tokens;
}

export function parse(src: string): Node {
  const tokens = tokenize(src);
  let pos = 0;
  const peek = () => tokens[pos];
  const next = () => tokens[pos++];
  const isOp = (v: string) => peek() !== undefined && peek().type === "op" && peek().value === v;

  const primary = (): Node => {
    const t = next();
    if (!t) throw new SyntaxError("Неожиданный конец условия");
    if (t.type === "ident") return { type: "field", name: t.value as string };
    if (t.type === "string" || t.type === "number" || t.type === "boolean") return { type: "literal", value: t.value };
    if (t.type === "op" && t.value === "(") {
      const e = or();
      if (!isOp(")")) throw new SyntaxError("Ожидалась закрывающая скобка");
      next();
      return e;
    }
    throw new SyntaxError(`Неожиданный токен «${String(t.value)}»`);
  };
  const compare = (): Node => {
    const left = primary();
    const t = peek();
    if (t && t.type === "op" && COMPARE_OPS.includes(t.value as string)) {
      const op = next().value as string;
      return { type: "compare", op, left, right: primary() };
    }
    return left;
  };
  const unary = (): Node => {
    if (isOp("!")) {
      next();
      return { type: "not", operand: unary() };
    }
    return compare();
  };
  const and = (): Node => {
    let left = unary();
    while (isOp("&&")) {
      next();
      left = { type: "and", left, right: unary() };
    }
    return left;
  };
  const or = (): Node => {
    let left = and();
    while (isOp("||")) {
      next();
      left = { type: "or", left, right: and() };
    }
    return left;
  };
  const ast = or();
  if (pos < tokens.length) throw new SyntaxError(`Лишний токен «${String(tokens[pos].value)}» после конца условия`);
  return ast;
}

export const truthy = (v: unknown): boolean => {
  if (Array.isArray(v)) return v.length > 0;
  if (typeof v === "string") return v !== "" && v.toLowerCase() !== "false";
  return Boolean(v);
};

const looseEquals = (a: unknown, b: unknown): boolean => {
  if (Array.isArray(a)) return a.some((x) => looseEquals(x, b));
  if (Array.isArray(b)) return b.some((x) => looseEquals(a, x));
  if (typeof a === "boolean" || typeof b === "boolean") return truthy(a) === truthy(b);
  if (a === undefined || a === null) a = "";
  if (b === undefined || b === null) b = "";
  return String(a) === String(b);
};

const numeric = (v: unknown): number | null => {
  const n = typeof v === "number" ? v : parseFloat(String(v).replace(",", "."));
  return Number.isNaN(n) ? null : n;
};

function evalNode(node: Node, values: Answers | undefined): unknown {
  switch (node.type) {
    case "field":
      return values ? values[node.name] : undefined;
    case "literal":
      return node.value;
    case "not":
      return !truthy(evalNode(node.operand, values));
    case "and":
      return truthy(evalNode(node.left, values)) && truthy(evalNode(node.right, values));
    case "or":
      return truthy(evalNode(node.left, values)) || truthy(evalNode(node.right, values));
    case "compare": {
      const l = evalNode(node.left, values);
      const r = evalNode(node.right, values);
      if (node.op === "==") return looseEquals(l, r);
      if (node.op === "!=") return !looseEquals(l, r);
      const ln = numeric(l);
      const rn = numeric(r);
      if (ln === null || rn === null) return false;
      if (node.op === ">") return ln > rn;
      if (node.op === ">=") return ln >= rn;
      if (node.op === "<") return ln < rn;
      return ln <= rn;
    }
  }
}

function fromObject(cond: Condition): Node {
  if (typeof cond === "string") return parse(cond);
  if (!cond || typeof cond !== "object") throw new SyntaxError("Неизвестная форма условия");
  if (Array.isArray(cond.any)) {
    return (
      cond.any.map(fromObject).reduce<Node | null>((acc, n) => (acc ? { type: "or", left: acc, right: n } : n), null) ?? {
        type: "literal",
        value: true,
      }
    );
  }
  if (Array.isArray(cond.all)) {
    return (
      cond.all.map(fromObject).reduce<Node | null>((acc, n) => (acc ? { type: "and", left: acc, right: n } : n), null) ?? {
        type: "literal",
        value: true,
      }
    );
  }
  if (cond.not !== undefined) return { type: "not", operand: fromObject(cond.not) };
  if (typeof cond.field === "string") {
    const op = cond.op || "==";
    if (!COMPARE_OPS.includes(op)) throw new SyntaxError(`Неизвестный оператор «${op}»`);
    return { type: "compare", op, left: { type: "field", name: cond.field }, right: { type: "literal", value: cond.value } };
  }
  throw new SyntaxError("Неизвестная форма условия");
}

const cache = new Map<string, Node>();

export function compile(condition: Condition): Node | null {
  if (condition === undefined || condition === null || condition === "") return null;
  if (typeof condition === "string") {
    if (condition.trim() === "") return null;
    let ast = cache.get(condition);
    if (!ast) {
      ast = parse(condition);
      cache.set(condition, ast);
    }
    return ast;
  }
  if (typeof condition === "object") return fromObject(condition);
  throw new SyntaxError("Условие должно быть строкой или объектом");
}

export function evaluate(condition: Condition, values: Answers | undefined, onError?: (err: Error, condition: Condition) => void): boolean {
  let ast: Node | null;
  try {
    ast = compile(condition);
  } catch (err) {
    if (onError) onError(err as Error, condition);
    return true;
  }
  if (ast === null) return true;
  return truthy(evalNode(ast, values));
}

/** Видимость элемента/компонента анкеты по его свойству conditions. */
export function isVisible(item: { conditions?: Condition } | null | undefined, values: Answers | undefined): boolean {
  if (!item || typeof item !== "object") return true;
  return evaluate(item.conditions, values, (err, cond) => {
    console.warn(`Некорректное условие показа «${String(cond)}»: ${err.message}. Элемент показан.`);
  });
}
